microceph/ceph/services.go

```diff
@@ -4,13 +4,132 @@ import (
 	"context"
 	"database/sql"
 	"fmt"
+	"time"
 
+	"github.com/Rican7/retry"
+	"github.com/Rican7/retry/backoff"
+	"github.com/Rican7/retry/strategy"
 	"github.com/canonical/microcluster/state"
+	"github.com/lxc/lxd/shared/logger"
 
 	"github.com/canonical/microceph/microceph/api/types"
 	"github.com/canonical/microceph/microceph/database"
+	"github.com/tidwall/gjson"
 )
 
+type Set map[string]struct{}
+
+func (sub Set) isIn(super Set) bool {
+	flag := true
+
+	// mark flag false if any key from subset is not present in superset.
+	for key := range sub {
+		_, ok := super[key]
+		if !ok {
+			flag = false
+			break // Break the loop.
+		}
+	}
+
+	return flag
+}
+
+// Table to map fetchFunc for workers (daemons) to a service.
+var serviceWorkerTable = map[string](func () (Set, error)) {
+	"osd": getUpOsds,
+	"mon": getMons,
+}
+
+// Restarts (in order) all Ceph Services provided in the input slice on the host.
+func RestartCephServices(services []string) error {
+	for i := range services {
+		err := RestartCephService(services[i])
+		if err != nil {
```

Comment:
If we check user input at the API boundary in the POST handler we could simplify the error handling here